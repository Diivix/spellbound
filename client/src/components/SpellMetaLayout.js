import React from 'react';
import PropTypes from 'prop-types';
import _ from 'lodash';
import {
  Grid,
  Header
} from 'semantic-ui-react';

class SpellMetaLayout extends React.Component {
  constructor(props) {
    super(props);

    this.buildCastingTimeValue = this.buildCastingTimeValue.bind(this);
    this.buildLevelWithSchool = this.buildLevelWithSchool.bind(this);
    this.buildRangeValue = this.buildRangeValue.bind(this);
  }


  buildCastingTimeValue = (castingTime, castingTimeDescription) => {
    if (castingTimeDescription) {
      castingTime = +' (' + castingTimeDescription + ')';
    }
    return castingTime;
  };

  buildRangeValue = (range, rangeDescription) => {
    if (rangeDescription) {
      range += ' (' + rangeDescription + ')';
    }
    return range;
  };

  buildLevelWithSchool = (level, school, truncateValue) => {
    school = _.upperFirst(school);

    let value;
    switch (level) {
      case 0:
        value = school + ' cantrip';
        break;
      case 1:
        value = level + 'st level ' + school;
        break;
      case 2:
        value = level + 'nd level ' + school;
        break;
      case 3:
        value = level + 'rd level ' + school;
        break;
      default:
        value = level + 'th level ' + school;
        break;
    }

    return (value = truncateValue ? _.truncate(value, { length: 20 }) : value);
  };

  render() {
    const levelWithSchool = this.buildLevelWithSchool(
      this.props.spell.level,
      this.props.spell.school,
      false
    );
    const components = this.props.spell.components
      .map(component => _.upperCase(component))
      .join(', ');
    const classes = this.props.spell.classes
      .map(clss => _.capitalize(clss))
      .join(' Â· ');
    const materials = _.upperFirst(this.props.spell.materials);
    const castingTime = this.props.spell.castingTimeDescription
      ? this.props.spell.castingTime +
      ' ' +
      this.props.spell.castingTimeDescription
      : this.props.spell.castingTime;
    const range = this.props.spell.rangeDescription
      ? this.props.spell.range + ' ' + this.props.spell.rangeDescription
      : this.props.spell.range;

    const duration = this.props.spell.durationDescription
      ? this.props.spell.duration + ' ' + this.props.spell.durationDescription
      : this.props.spell.duration;

    return (
      <Grid celled="internally">
        <Grid.Row columns="1" textAlign="center">
          <Grid.Column>
            <div>
              <Header sub color="grey" size="tiny">
                {this.props.spell.name}
              </Header>
              <div>
                <i>{levelWithSchool}</i>
              </div>
            </div>
          </Grid.Column>
        </Grid.Row>

        <Grid.Row columns="2" textAlign="center">
          <Grid.Column>
            <div>
              <Header sub color="grey" size="tiny">
                Casting Time
                </Header>
              <div>{castingTime}</div>
            </div>
          </Grid.Column>
          <Grid.Column>
            <div>
              <Header sub color="grey" size="tiny">
                Range
                </Header>
              <div>{range}</div>
            </div>
          </Grid.Column>
        </Grid.Row>

        <Grid.Row columns="2" textAlign="center">
          <Grid.Column>
            <div>
              <Header sub color="grey" size="tiny">
                Components
                </Header>
              <div>
                {_.isEmpty(components) ? '-' : <span>{components}</span>}
              </div>
            </div>
          </Grid.Column>
          <Grid.Column>
            <div>
              <Header sub color="grey" size="tiny">
                Duration
                </Header>
              <div>{duration}</div>
            </div>
          </Grid.Column>
        </Grid.Row>

        {_.isEmpty(materials) ? (
          ''
        ) : (
            <Grid.Row columns="1">
              <Grid.Column>
                <div style={{ textAlign: 'justify' }}>
                  <div>
                    <i>{materials}</i>
                  </div>
                </div>
              </Grid.Column>
            </Grid.Row>
          )}

        <Grid.Row columns="1" textAlign="center">
          <Grid.Column>
            <div style={{ color: 'grey' }}>
              <i>{classes}</i>
            </div>
          </Grid.Column>
        </Grid.Row>
      </Grid>
    );

  }
}

SpellMetaLayout.propTypes = {
  spell: PropTypes.shape({
    _id: PropTypes.string.isRequired,
    castingTime: PropTypes.string.isRequired,
    castingTimeDescription: PropTypes.string.isRequired,
    classes: PropTypes.arrayOf(PropTypes.string).isRequired,
    components: PropTypes.arrayOf(PropTypes.string).isRequired,
    duration: PropTypes.string.isRequired,
    durationDescription: PropTypes.string.isRequired,
    level: PropTypes.number.isRequired,
    name: PropTypes.string.isRequired,
    range: PropTypes.string.isRequired,
    rangeDescription: PropTypes.string.isRequired,
    school: PropTypes.string.isRequired
  }).isRequired
};

export default SpellMetaLayout;