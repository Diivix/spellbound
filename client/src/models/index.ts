export interface ISpell {
  _id: string;
  atHigherLevels: string;
  castingTime: string;
  castingTimeDescription: string;
  classes: string[];
  components: string[];
  description: string;
  duration: string;
  durationDescription: string;
  level: number;
  name: string;
  range: string;
  rangeDescription: string;
  school: string;
}

export interface ISpellCompendiumState {
  // filters: IFilters;
  spells: ISpell[];
}

export interface IAuthState {
  authStatus: string;
}

export interface ICredentials {
  email: string;
  password: string;
}