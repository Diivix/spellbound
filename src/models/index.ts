import { RouterState } from 'react-router-redux';

export interface IStoreState {
  isAuthenticated: boolean;
  pendingActions: number;
  router?: RouterState;
  spellData: ISpellData;
  userData: IUserData;
}

export interface IUserData {
  token: string;
  userName: string;
  characters: ICharacter[];
}

export interface ISpellData {
  currentSpell?: ISpell;
  spells?: ISpell[];
  filters?: IFilters;
  appliedFilters?: IFilters;
}

export interface ISpell {
  _id: string;
  atHigherLevels?: string;
  castingTime: string;
  classTypes: string[];
  components: string[];
  description?: string;
  duration: string;
  level: number;
  materials?: string;
  name: string;
  range: string;
  reference: string;
  school: string;
}

export interface ICharacterBase {
  name: string;
  level?: number;
  classType?: string;
  description?: string;
}
export interface ICharacter extends ICharacterBase {
  _id: string;
  dateCreated: number;
  dateLastModified: number;
  spells?: ISpell[];
}

export interface IFilters {
  classTypes: ISelectItem[];
  components: ISelectItem[];
  names: ISelectItem[];
  ranges: ISelectItem[];
  schools: ISelectItem[];
}

export interface ICredentials {
  email: string;
  password: string;
}

export interface ISelectItem {
  key: string;
  value: string;
}
