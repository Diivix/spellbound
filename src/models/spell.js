import mongoose from 'mongoose';

var spells = new mongoose.Schema({  
  name: { type: String, lowercase: true },
  school: { type: String, lowercase: true },
  level: Number,
  classes: [{ type: String, lowercase: true }],
  castingTime: { type: String, lowercase: true },
  castingTimeDescription: { type: String, lowercase: true },
  range: { type: String, lowercase: true },
  rangeDescription: { type: String, lowercase: true },
  components: [{ type: String, lowercase: true }],
  duration: { type: String, lowercase: true },
  durationDescription: { type: String, lowercase: true },
  description: String,
  atHigherLevels: String
});
mongoose.model('spell', spells);

export default mongoose.model('spell');