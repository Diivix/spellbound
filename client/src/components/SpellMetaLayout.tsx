import _ from 'lodash';
import React from 'react';
import { Grid, Header } from 'semantic-ui-react';
import { ISpell } from '../models';

interface IProps {
  spell: ISpell;
}

class SpellMetaLayoutComponent extends React.Component<IProps, {}> {
  constructor(props: IProps) {
    super(props);
  }

  public buildCastingTimeValue = (castingTime: string, castingTimeDescription: string) => {
    if (castingTimeDescription) {
      castingTime = +' (' + castingTimeDescription + ')';
    }
    return castingTime;
  };

  public buildRangeValue = (range: string, rangeDescription: string) => {
    if (rangeDescription) {
      range += ' (' + rangeDescription + ')';
    }
    return range;
  };

  public buildLevelWithSchool = (level: number, school: string, truncateValue: boolean) => {
    school = _.upperFirst(school);

    let value;
    switch (level) {
      case 0:
        value = school + ' cantrip';
        break;
      case 1:
        value = level + 'st level ' + school;
        break;
      case 2:
        value = level + 'nd level ' + school;
        break;
      case 3:
        value = level + 'rd level ' + school;
        break;
      default:
        value = level + 'th level ' + school;
        break;
    }

    return (value = truncateValue ? _.truncate(value, { length: 20 }) : value);
  };

  public render() {
    const levelWithSchool = this.buildLevelWithSchool(this.props.spell.level, this.props.spell.school, false);
    const components = this.props.spell.components.map(component => _.upperCase(component)).join(', ');

    const classes = this.props.spell.classes.map(clss => _.capitalize(clss)).join(' Â· ');

    const materials = _.upperFirst(this.props.spell.materials);

    const castingTime = this.props.spell.castingTimeDescription
      ? this.props.spell.castingTime + ' ' + this.props.spell.castingTimeDescription
      : this.props.spell.castingTime;

    const range = this.props.spell.rangeDescription
      ? this.props.spell.range + ' ' + this.props.spell.rangeDescription
      : this.props.spell.range;

    const duration = this.props.spell.durationDescription
      ? this.props.spell.duration + ' ' + this.props.spell.durationDescription
      : this.props.spell.duration;

    return (
      <Grid celled="internally">
        <Grid.Row columns="1" textAlign="center">
          <Grid.Column>
            <div>
              <Header sub={true} color="grey" size="tiny">
                {this.props.spell.name}
              </Header>
              <div>
                <i>{levelWithSchool}</i>
              </div>
            </div>
          </Grid.Column>
        </Grid.Row>

        <Grid.Row columns="2" textAlign="center">
          <Grid.Column>
            <div>
              <Header sub={true} color="grey" size="tiny">
                Casting Time
              </Header>
              <div>{castingTime}</div>
            </div>
          </Grid.Column>
          <Grid.Column>
            <div>
              <Header sub={true} color="grey" size="tiny">
                Range
              </Header>
              <div>{range}</div>
            </div>
          </Grid.Column>
        </Grid.Row>

        <Grid.Row columns="2" textAlign="center">
          <Grid.Column>
            <div>
              <Header sub={true} color="grey" size="tiny">
                Components
              </Header>
              <div>{_.isEmpty(components) ? '-' : <span>{components}</span>}</div>
            </div>
          </Grid.Column>
          <Grid.Column>
            <div>
              <Header sub={true} color="grey" size="tiny">
                Duration
              </Header>
              <div>{duration}</div>
            </div>
          </Grid.Column>
        </Grid.Row>

        {_.isEmpty(materials) ? (
          ''
        ) : (
          <Grid.Row columns="1">
            <Grid.Column>
              <div style={{ textAlign: 'justify' }}>
                <div>
                  <i>{materials}</i>
                </div>
              </div>
            </Grid.Column>
          </Grid.Row>
        )}

        <Grid.Row columns="1" textAlign="center">
          <Grid.Column>
            <div style={{ color: 'grey' }}>
              <i>{classes}</i>
            </div>
          </Grid.Column>
        </Grid.Row>
      </Grid>
    );
  }
}

export default SpellMetaLayoutComponent;
