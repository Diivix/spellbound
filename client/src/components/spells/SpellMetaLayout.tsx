import _ from 'lodash';
import React from 'react';
import { Grid, Header } from 'semantic-ui-react';
import { BuildLevelWithSchool } from 'utils/ui';
import { ISpell } from '../../models';

interface IProps {
  spell: ISpell;
}

class SpellMetaLayoutComponent extends React.Component<IProps, {}> {
  constructor(props: IProps) {
    super(props);
  }

  public render() {
    const { spell } = this.props;
    const levelWithSchool = BuildLevelWithSchool(spell.level, spell.school, false);
    const components = spell.components.map(component => _.upperCase(component)).join(', ');
    const classTypes = spell.classTypes.map(clss => _.capitalize(clss)).join(' Â· ');
    const castingTime = spell.castingTime;
    const range = spell.range;
    const duration = spell.duration;

    let materialElement = <div />;
    if (!_.isEmpty(spell.materials)) {
      materialElement = (
        <Grid.Row columns="1">
          <Grid.Column textAlign="center">
            <i>{_.upperFirst(spell.materials)}.</i>
          </Grid.Column>
        </Grid.Row>
      );
    }

    return (
      <Grid celled="internally">
        <Grid.Row columns="1" textAlign="center">
          <Grid.Column>
            <Header sub={true} color="grey" size="tiny">
              {spell.name}
            </Header>
            <i>{levelWithSchool}</i>
          </Grid.Column>
        </Grid.Row>

        <Grid.Row columns="2" textAlign="center">
          <Grid.Column>
            <Header sub={true} color="grey" size="tiny">
              Casting Time
            </Header>
            {castingTime}
          </Grid.Column>
          <Grid.Column>
            <Header sub={true} color="grey" size="tiny">
              Range
            </Header>
            {range}
          </Grid.Column>
        </Grid.Row>

        <Grid.Row columns="2" textAlign="center">
          <Grid.Column>
            <Header sub={true} color="grey" size="tiny">
              Components
            </Header>
            {_.isEmpty(components) ? '-' : <span>{components}</span>}
          </Grid.Column>
          <Grid.Column>
            <Header sub={true} color="grey" size="tiny">
              Duration
            </Header>
            {duration}
          </Grid.Column>
        </Grid.Row>

        {materialElement}

        <Grid.Row columns="1" textAlign="center">
          <Grid.Column>
            <div style={{ color: 'grey' }}>
              <i>{classTypes}</i>
            </div>
          </Grid.Column>
        </Grid.Row>
      </Grid>
    );
  }
}

export default SpellMetaLayoutComponent;
