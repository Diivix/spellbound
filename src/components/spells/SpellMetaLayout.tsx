import _ from 'lodash';
import React from 'react';
import { BuildLevelWithSchool } from 'utils/ui';
import { ISpell } from '../../models';

interface IProps {
  spell: ISpell;
}

class SpellMetaLayoutComponent extends React.Component<IProps, {}> {
  constructor(props: IProps) {
    super(props);
  }

  public render() {
    const { spell } = this.props;
    const levelWithSchool = BuildLevelWithSchool(spell.level, spell.school, false);
    const components = spell.components.map(component => _.upperCase(component)).join(', ');
    const classTypes = spell.classTypes.map(clss => _.capitalize(clss)).join(' Â· ');
    const castingTime = spell.castingTime;
    const range = spell.range;
    const duration = spell.duration;

    let materialElement = <div />;
    if (!_.isEmpty(spell.materials)) {
      materialElement = (
        <div className="sb-row">
          <div className="sb-col">
            <p className="center">
              <i>{_.upperFirst(spell.materials)}.</i>
            </p>
          </div>
        </div>
      );
    }

    return (
      <div className="sb-physical-card">
        <div className="sb-row">
          <div className="sb-col">
            <h3>{spell.name}</h3>
            <i>{levelWithSchool}</i>
          </div>
        </div>

        <div className="sb-row">
          <div className="sb-col">
            <h3>Casting Time</h3>
            {castingTime}
          </div>
          <div className="sb-col">
            <h3>Range</h3>
            {range}
          </div>
        </div>

        <div className="sb-row">
          <div className="sb-col">
            <h3>Components</h3>
            {_.isEmpty(components) ? '-' : <span>{components}</span>}
          </div>
          <div className="sb-col">
            <h3>Duration</h3>
            {duration}
          </div>
        </div>

        {materialElement}

        <div className="sb-row">
          <div className="sb-col">
            <p className="grey">
              <i>{classTypes}</i>
            </p>
          </div>
        </div>
      </div>
    );
  }
}

export default SpellMetaLayoutComponent;
