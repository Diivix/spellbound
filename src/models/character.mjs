import { spells as spellsSchema } from './spell';
import mongoose from 'mongoose';

export const characters = new mongoose.Schema({
    name: {
        type: String,
        required: true,
        trim: true
    },
    level: {
        type: Number,
        default: 0
    },
    class: {
        type: String,
        trime: true
    },
    description: {
        type: String,
        trim: true
    },
    dateCreated: {
        type: Date
    },
    dateLastModified: {
        type: Date,
        default: Date.now
    },
    spells: [spellsSchema]
});
