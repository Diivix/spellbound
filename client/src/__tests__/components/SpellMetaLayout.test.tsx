import React from 'react';
import renderer from 'react-test-renderer';
import SpellMetaLayout from '../../components/spells/SpellMetaLayout';
import { ISpell } from '../../models';

test('Show Spell Metadata', () => {
  const spell: ISpell = {
    _id: "0",
    atHigherLevels: '',
    castingTime: '1',
    castingTimeDescription: 'Action',
    classTypes: ['Sorcerer', 'Wizard'],
    components: ['V', 'S'],
    description: '',
    duration: 'action',
    durationDescription: '',
    level: 6,
    name: 'Fire Bolt',
    range: '20',
    rangeDescription: 'feet',
    school: 'Destruction'
  }

  const component = renderer.create(
    <SpellMetaLayout spell={spell} />,
  );
  
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});