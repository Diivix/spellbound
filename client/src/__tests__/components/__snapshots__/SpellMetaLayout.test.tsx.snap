// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Show Spell Metadata 1`] = `
<div
  className="ui internally celled grid"
>
  <div
    className="center aligned one column row"
  >
    <div
      className="column"
    >
      <div
        className="ui grey tiny sub header"
      >
        Fire Bolt
      </div>
      <i>
        6th level Destruction
      </i>
    </div>
  </div>
  <div
    className="center aligned two column row"
  >
    <div
      className="column"
    >
      <div
        className="ui grey tiny sub header"
      >
        Casting Time
      </div>
      1 Action
    </div>
    <div
      className="column"
    >
      <div
        className="ui grey tiny sub header"
      >
        Range
      </div>
      20 feet
    </div>
  </div>
  <div
    className="center aligned two column row"
  >
    <div
      className="column"
    >
      <div
        className="ui grey tiny sub header"
      >
        Components
      </div>
      <span>
        V, S
      </span>
    </div>
    <div
      className="column"
    >
      <div
        className="ui grey tiny sub header"
      >
        Duration
      </div>
      action
    </div>
  </div>
  <div />
  <div
    className="center aligned one column row"
  >
    <div
      className="column"
    >
      <div
        style={
          Object {
            "color": "grey",
          }
        }
      >
        <i>
          Sorcerer Â· Wizard
        </i>
      </div>
    </div>
  </div>
</div>
`;
